### PR TITLE
Refactor arguments and options parsing

### LINKED ISSUE #813: this.option -> type doesn't coerce
So I would expect that by setting an option to a type, that any values passed via that option would at least get coerced.  Instead what happens is the type turns into whatever the passed in value seems to be (didn't look at the code for how a type is figured).

For example

```
git clone https://github.com/olsonpm/generator-option-type
cd generator-option-type
npm install
npm link

# now make a temporary directory - assume the below doesn't exist
mkdir -p ~/tmp.d/yeoman
cd ~/tmp.d/yeoman

# typeof aBoolOpt is undefined - makes sense since we never passed it
yo option-type
> typeof aBoolOpt: undefined

# typeof aBoolOpt is a boolean - also expected
yo option-type --aBoolOpt
> typeof aBoolOpt: boolean

# typeof aBoolOpt is a string, then a number
yo option-type --aBoolOpt=false
> typeof aBoolOpt: string

yo option-type --aBoolOpt=0
> typeof aBoolOpt: number
```

If this is the behavior, then what exactly is the point of declaring an option type?

### LINKED ISSUE #821: Use `minimist` instead of `nopt`
To avoid conflicts with `yo` (which uses `minimist`) I think it would be best to use `minimist` here too. This doesn't come without some issues though. By default, `nopt` treats `--foo bar` as `foo: true` (which is really awkward) while `minimist` would parse it as `foo: bar`, if we haven't told it that `foo` is a boolean.

If we're ok with dropping the old behavior I think it would fix issues like https://github.com/yeoman/yo/issues/319 and potential future issues.

### PR DESCRIPTION
**Pull request by SBoudrias**
Now use minimist to match yo and fix type parsing.

Fix #821 (use minimist instead of nopt)
Fix #813 (correct type casting)

This also consolidate all arguments defined in a Generator to `this.options.argName` (no more `this.argName`)

I'd appreciate some review given this is a pretty major change to our core system. cc @kevva @arthurvr @mischah @zckrs @olsonpm @binduwavell

### CODE REVIEW

**Comment by olsonpm** [comment]
I will try to test this sometime this week.  I'm impressed you tackled the problem - I remember it affecting a decent portion of the codebase.

**Comment by mischah** [comment]
Code-wise it looks good as far as I can tell. Going to prepare and execute a few manual test later this evening.

File: package.json [hunk 1]
```diff
@@ -75,6 +74,7 @@
     "gulp-plumber": "^1.0.0",
     "inquirer": "^1.0.2",
     "jsdoc": "^3.3.0-beta1",
+    "minimist": "^1.2.0",
```
Comment (mischah) [inline package.json]: `minimist` should be a dependency instead of a devDependency, right?

**Comment by mischah** [comment]
I did some manual tests with the following setup:

1. Generated a generator with generator-generator 😆
2. Replaced `"yeoman-generator": "^0.23.0",` with `"yeoman-generator": "yeoman/generator#feature/refactor-opts-parsing",` in package.json
3. Linked the generator with `npm link`

## Results:

1. ✅ Default options like `--help` and `--skip-install` 
2. ✅  Own options like `--coffee` decribed [here](http://yeoman.io/authoring/user-interactions.html)
3. ❌  Arguments like described [here](http://yeoman.io/authoring/user-interactions.html)

### Regarding 3.

I have used the following code in the constructor:
```js
  constructor: function () {
    yeoman.Base.apply(this, arguments);

    this.argument('appname', {
      type: String,
      default: 'My Appchen'
    });

    console.log('this.appname =', this.appname);
  }
```

**With yeoman-generator `0.23.4` it behaves like one would expect:**

```bash
$ yo temp --skip-install --coffee
this.appname = My Appchen
```

```bash
$ yo temp huhu --skip-install --coffee
this.appname = huhu
```

**With yeoman-generator from cf87a1b it behaves quite different:**

```bash
$ yo temp --skip-install --coffee
this.appname = test generator temp
```

```bash
$ yo temp huhu --skip-install --coffee
this.appname = test generator temp
```

… which is my working dir without the hyphens:
```bash
$ pwd
/Users/mkuehnel/temp/test-generator-temp
```

----

Another thing that really confuses me, is that the Base class is gone 😕 
So I had to do:

```js
module.exports = yeoman.extend({
…
})
```

Instead of:

```js
module.exports = yeoman.Base.extend({
…
})
```

The Base class is also not in the generated Docs. Did I slept on something?

**Comment by olsonpm** [comment]
Just got around to testing the same scenarios I raised in #813 they all pass as expected.  I only tripped up on not being able to write es6 (I thought `constructor() { /* do stuff */ }` would behave the same as `constructor: function constructor() { /* do stuff */ }` which it doesn't.)

**Comment by SBoudrias** [comment]
@olsonpm it should work the same - it does for me: https://github.com/SBoudrias/generator-jest/blob/master/generators/app/index.js

**Comment by olsonpm** [comment]
ha thanks - didn't realize I had to go full es6 (haven't played around with class yet).

This is what didn't work for me:
https://gist.github.com/olsonpm/c71d8cad6a0a219cea11e84aed8723db

**Comment by SBoudrias** [comment]
I would've assume your structure would've worked. I don't know what went wrong there.

**Comment by olsonpm** [comment]
Yeah - didn't mean to take any of your time.  It's a negligible problem :)

**Pull request closed without merging**